From: Justin Maggard <jmaggard10@gmail.com>
Date: Thu, 24 Sep 2020 08:55:36 -0700
Subject: upnphttp: Disallow negative HTTP chunk lengths

[CVE-2020-28926]

This fixes a couple vulnerabilities that could lead to an infinite loop
or heap corruption.
---
 upnphttp.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/upnphttp.c b/upnphttp.c
index 3b4b58a..61864e3 100644
--- a/upnphttp.c
+++ b/upnphttp.c
@@ -415,14 +415,14 @@ next_header:
 			return;
 		line += 2;
 	}
-	if( h->reqflags & FLAG_CHUNKED )
+	if (h->reqflags & FLAG_CHUNKED)
 	{
 		char *endptr;
 		h->req_chunklen = -1;
-		if( h->req_buflen <= h->req_contentoff )
+		if (h->req_buflen <= h->req_contentoff)
 			return;
 		while( (line < (h->req_buf + h->req_buflen)) &&
-		       (h->req_chunklen = strtol(line, &endptr, 16)) &&
+		       (h->req_chunklen = strtol(line, &endptr, 16) > 0) &&
 		       (endptr != line) )
 		{
 			endptr = strstr(endptr, "\r\n");
@@ -866,7 +866,7 @@ ProcessHttpQuery_upnphttp(struct upnphttp * h)
 		char *chunkstart, *chunk, *endptr, *endbuf;
 		chunk = endbuf = chunkstart = h->req_buf + h->req_contentoff;
 
-		while( (h->req_chunklen = strtol(chunk, &endptr, 16)) && (endptr != chunk) )
+		while ((h->req_chunklen = strtol(chunk, &endptr, 16)) > 0 && (endptr != chunk) )
 		{
 			endptr = strstr(endptr, "\r\n");
 			if (!endptr)
